waiter/src/waiter/token.clj

```diff
@@ -719,3 +725,119 @@
                                              :status http-405-method-not-allowed})))
     (catch Exception ex
       (utils/exception->response ex req))))
+
+(defn make-index-event
+  [type object]
+  {:object object :type type})
+
+(defn start-tokens-watch-maintainer
+  [kv-store tokens-update-chan tokens-watch-channels-update-chan watch-refresh-timeout-ms exit-chan]
+  (let [tokens-event-chan (au/latest-chan)
+        tokens-event-mult (async/mult tokens-event-chan)
+        query-chan (async/chan)
+        get-token->token-index-fn (fn []
+                                    (->> kv-store
+                                         list-token-owners
+                                         (reduce
+                                           (fn [current-token-index-map owner]
+                                             (reduce
+                                               (fn [inner-token-index-map [token entry]]
+                                                 (->> (assoc entry :owner owner :token token)
+                                                      (assoc inner-token-index-map token)))
+                                               current-token-index-map
+                                               (list-index-entries-for-owner kv-store owner :refresh true)))
+                                           {})))
+        state-atom (atom {:token->token-index (get-token->token-index-fn)
+                          :watches-count 0})
+        go-chan
+        (async/go
+          (try
+            (loop [{:keys [token->token-index watches-count] :as current-state} @state-atom
+                   watch-refresh-timeout-chan (async/timeout watch-refresh-timeout-ms)]
+              (reset! state-atom current-state)
+              (let [[msg current-chan]
+                    (async/alts! [exit-chan tokens-update-chan tokens-watch-channels-update-chan
+                                  watch-refresh-timeout-chan query-chan]
+                                 :priority true)
+                    [next-state next-watch-refresh-timeout-chan]
+                    (condp = current-chan
+                      exit-chan
+                      (do
+                        (log/warn "Stopping tokens-watch-maintainer")
+                        (when (not= :exit msg)
+                          (throw (ex-info "Stopping router-state maintainer" {:time (t/now) :reason msg}))))
+
+                      tokens-update-chan
+                      (timers/start-stop-time!
+                        (metrics/waiter-timer "core" "tokens-watch-maintainer" "process-tokens-update")
+                        (let [{:keys [token owner]} msg
+                              token-index-entry (some-> (list-index-entries-for-owner kv-store owner :refresh true)
```

Comment:
Can we have a helper function, `token->index-entry`, instead? It might be useful in `get-token->token-index-fn` also.